### ISSUE TITLE
Does this library work on MacOSX?

### ISSUE THREAD

**Issue by Amaruk2365** [pfroud/JVisa#3]
Thanks for sharing this fantastic library!!! I am trying out the SimpleExample you have here but I get run-time errors (see below). I am using this on MacOSX but I did not do any changes to the code (it seems to point to a VISA .dll). Any help is appreciated! Thank you again!
Details: MacOS Catalina version 10.15.7
The computer has LabVIEW 2021 installed and I can control instruments just fine with VISA from within LabVIEW.

`Exception in thread "main" java.lang.NoClassDefFoundError: com/sun/jna/Library
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1012)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:862)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:760)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:681)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:639)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at jvisa.JVisaResourceManager.<init>(JVisaResourceManager.java:58)`

**Comment by pfroud** [comment]
I appreciate your appreciation!

It looks like JVisa should be compatible with macOS. The first step is to fix the `NoClassDefFoundError`, which means Java can't find [JNA](https://github.com/java-native-access/jna). JNA is the library which lets JVisa talk to NI-VISA.  

I suggest using [NetBeans IDE](https://netbeans.apache.org/) for testing this. The JVisa git reposority contains files which make it a Netbeans project. If you clone the git reposority then go to Netbeans and click File > Open Project, then Netbeans should recognize the JVisa folder as a Netbeans project. The git reposority also contains a compiled JNA jar file, and the Netbeans project is already configured up to use it.

You are right that JVisa currently points to a `dll` file, here's the line:
https://github.com/pfroud/JVisa/blob/a018392f831c9b4eb2ade86e2c39417d46294f5d/src/jvisa/JVisaResourceManager.java#L58

On a Mac, it looks like the file extension will be `.dylib` (dynamic library) or maybe `.so` (shared object). The file name could be `libvisa` or `libnivisa`. You can read the section called "Installed Files and Paths" in the [NI-VISA 21.0 for macOS Readme](https://www.ni.com/pdf/manuals/375354j.html#installation) to see where to look for it.



If you get it working, we can update the Java code to check the OS type.

**Comment by Amaruk2365** [comment]
Thank you for your quick reply!!!
Ahhh... I had not imported PNA into this project. I thought it was part of JVisa... I have tested my JNA library before in a separate project and I got that example to work (but it had no VISA functions).

After also adding JNA to my project I got the error that I we were expecting (.dll file) as shown below:

```
Couldn't load nivisa.dll
java.lang.UnsatisfiedLinkError: Unable to load library 'nivisa64.dll':
dlopen(libnivisa64.dll.dylib, 9): image not found
dlopen(libnivisa64.dll.dylib, 9): image not found
dlopen(/Users/mikaelgarcia/Library/Frameworks/nivisa64.dll.framework/nivisa64.dll, 9): image not found
dlopen(/Library/Frameworks/nivisa64.dll.framework/nivisa64.dll, 9): image not found
dlopen(/System/Library/Frameworks/nivisa64.dll.framework/nivisa64.dll, 9): image not found
```

In your JVisa code there is this section:

```
public class JVisaResourceManager {
    private final NativeLong RESOURCE_MANAGER_HANDLE;
    public final JVisaLibrary VISA_LIBRARY = (JVisaLibrary)Native.load("nivisa64.dll", JVisaLibrary.class
```
My issue now is that I don't know how to modify your library as I get a read-only error if I try to edit the file name listed above... I have never modified a library file before so if you could give me a hint on how to do this that would be great! Thanks again! I should also say that I am using IntelliJ IDE for this project. Thanks gain!

**Comment by pfroud** [comment]
OK good progress.

There's nothing special about these .java files, you should be able to edit them. How did you get the JVisa code on your computer? (Cloned using command-line git? Downloaded a ZIP from github.com?)

**Comment by Amaruk2365** [comment]
Hi Peter,
I downloaded the .jar file directly from GitHub (JVisa-2022-01-04.jar). And
I just linked to it in my test project. But on a Mac I can't just edit the
.jar file and in my IDE I also don't seem to be able to write to it either.
I am new to Java so perhaps I should have done something different (just
download a .java file??). I am very thankful for any help. I have many
years experience with NI products (LabVIEW, NI-MAX etc) and I have also
done a lot of instrument control using VISA in C++ on Windows 7/10 in the
past. But this time I would like to use Java and since NI does not support
Java with their VISA your project with JVisa fills a big void and would be
fantastic to have! Hoping we can figure this out. If you want to try it out
on Linux too I can install a Java IDE on my Linux PC and run the same tests
once we have the Mac up and running.

Thanks again!!

Best regards,
Mikael

**Comment by pfroud** [comment]
For this you'll need to get the JVisa source code, which are .java files.

A jar file is a package of _compiled_ java classes. A jar lets you _use_ the contents, but not edit them. Confusingly, your IDE tries to be helpful by automatically decompiling the jar file to show you source code -- but you can't edit it.

To get the JVisa source code you can click this link: https://github.com/pfroud/JVisa/archive/refs/heads/master.zip
It will download a file called JVisa-master.zip. Extract the zip file to a folder called JVisa-master.

NetBeans IDE will recognize the folder JVisa-master as a Netbeans project.  It looks like IntelliJ will also be able to open the folder.

**Comment by Amaruk2365** [comment]
Hi Peter,
Thank you very much!!I am going to give this a try. Will let you know how
it goes. Thanks again!
Best regards,
Mikael

**Comment by Amaruk2365** [comment]
Hi again Peter,

I was able to download the zip file, unzip it and have IntelliJ recognize
the project folder! Great step forward!! I am now able to edit the source
code but I have not done any changes yet.

Below you can see the code as it shows up (key section) and the error
messages I get. What do you suggest I change the VISA_LIBRARY value to? It
seems your code adds the extension automatically now. Perhaps I need to
specify where I have my VISA libraries installed on my Mac? Let me know
what to look for and I can see where they reside on my Mac.

Best regards,
Mikael

Here is the dll section in your master project that I used this time which
I have left as is:

/*

You do NOT need to specify the file extension.

JNA does a bunch of checks and might use java.lang.System.mapLibraryName().

You can see what JNA is doing here:

https://github.com/java-native-access/jna/blob/master/src/com/sun/jna/NativeLibrary.java

Here's a helpful StackOverflow answer:

https://stackoverflow.com/a/37329511/7376577

 */

final String nativeLibraryName = "nivisa64";

VISA_LIBRARY = (JVisaLibrary) Native.load(nativeLibraryName,
JVisaLibrary.class);


Here are the run-time error messages I get when I run it.


Couldn't load nivisa.dll

java.lang.UnsatisfiedLinkError: Unable to load library 'nivisa64':

dlopen(libnivisa64.dylib, 9): image not found

dlopen(libnivisa64.dylib, 9): image not found

Native library (darwin/libnivisa64.dylib) not found in resource path (/Users/..

**Comment by pfroud** [comment]
Now you need to go looking for a dylib file that looks like belongs to Visa.

According to this page https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/UsingDynamicLibraries.html

> The standard locations for dynamic libraries are `~/lib`, `/usr/local/lib`, and `/usr/lib`.

**Comment by Amaruk2365** [comment]
Hi Peter,
I don't know much about these libraries but it seems that NI is using the
frameworks concept mentioned here:
https://www.ni.com/pdf/manuals/375354h.html
I did however find the VISA.framework folder so it is on my system. But no
luck in finding any dylib files there yet (I am using the terminal window
for this and I am no pro in using advanced Unix commands so it is a very
primitive search I have done so far so it might be there but I have not
found it yet). I have also not found any documentation talking about the
dylib files in this folder. Perhaps they are using things in a different
manner? Any ideas of what to look for and where to look - I am all ears.
Thanks again!
Best regards,
Mikael

**Comment by pfroud** [comment]
In JVisaResourceManager.java, try setting `String nativeLibraryName` to `visa`, `visa32`, and `visa64` then try running SimpleExample.java each time

**Comment by Amaruk2365** [comment]
We are getting closer...
When I set it to just "visa" (see below) I don't get the same error message
which seems encouraging! Here is the declaration and the run-time message I
get now. I should point out that I have an instrument connected on the same
network (LAN) that works fine when I ping it (=I can talk to it from my
other programs). I did not have any USB or GPIB devices connected for this
test.

Thanks again!
Mikael

*final *String nativeLibraryName = *"visa"*;



Couldn't find any VISA resources

jvisa.JVisaException: in viFindRsrc(): Insufficient location information or
the device or resource is not present in the system. (0xBFFF0011)

at jvisa.JVisaUtils.throwForStatus(JVisaUtils.java:61)

at jvisa.JVisaResourceManager.findResources(JVisaResourceManager.java:171)

at jvisa_example.lowlevel.SimpleExample.main(SimpleExample.java:60)


Process finished with exit code 0

**Comment by pfroud** [comment]
Excellent news!

I hard-coded JVisa to only search for USB instruments which was dumb.

Please open JVisaResourceManager.java and edit line 159:

```
 final ByteBuffer filterExpression = stringToByteBuffer("USB?*");
```
Change the string from `"USB?*"` to `"?*"`, so the line is
```
 final ByteBuffer filterExpression = stringToByteBuffer("?*");
```
And then run Simple Example again.

If everything works I will publish another release on Github.

**Comment by Amaruk2365** [comment]
Hi Peter,

We are almost there!!

I changed the line to "?*" as you suggested and ran two experiments. See
details below.

Thanks again!!

Best regards,
Mikael

Experiment 1.
Instruments connected:

   - Keithley 236 SMU on GPIB address 10 (old, does not support *IDN?
   commands as you will see below "NSDCV+...")
   - Keithley 236 SMU on GPIB address 12 (old, does not support *IDN?
   commands as you will see below "NSDCV+...")
   - Siglent SDM3045X DMM on LAN address 10.0.1.21


This is the output I get:

Found 6 VISA instrument(s):
jvisa.JVisaException: in viWrite(): Unable to start operation because setup
is invalid (due to attributes being set to an inconsistent state).
(0xBFFF003A)
at jvisa.JVisaUtils.throwForStatus(JVisaUtils.java:61)
at jvisa.JVisaInstrument.write(JVisaInstrument.java:119)
at jvisa.JVisaInstrument.sendAndReceiveString(JVisaInstrument.java:75)
at jvisa_example.lowlevel.SimpleExample.main(SimpleExample.java:76)
2 / 6: "NSDCV+0.0000E+00,B0000"
3 / 6: "NSDCV+0.0000E+00,B0000"
4 / 6: "Siglent Technologies,SDM3045X,SDM34FBD4R1838,5.01.01.06R1"
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)
at jvisa.JVisaUtils.throwForStatus(JVisaUtils.java:61)
at jvisa.JVisaResourceManager.openInstrument(JVisaResourceManager.java:143)
at jvisa_example.lowlevel.SimpleExample.main(SimpleExample.java:75)
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)
at jvisa.JVisaUtils.throwForStatus(JVisaUtils.java:61)
at jvisa.JVisaResourceManager.openInstrument(JVisaResourceManager.java:143)
at jvisa_example.lowlevel.SimpleExample.main(SimpleExample.java:75)



Experiment 2.
Instruments connected:

   - Siglent SDM3045X DMM on LAN address 10.0.1.21

This is the output I get:


Found 3 VISA instrument(s):
1 / 3: "Siglent Technologies,SDM3045X,SDM34FBD4R1838,5.01.01.06R1"
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)
at jvisa.JVisaUtils.throwForStatus(JVisaUtils.java:61)
at jvisa.JVisaResourceManager.openInstrument(JVisaResourceManager.java:143)
at jvisa_example.lowlevel.SimpleExample.main(SimpleExample.java:75)
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)
at jvisa.JVisaUtils.throwForStatus(JVisaUtils.java:61)
at jvisa.JVisaResourceManager.openInstrument(JVisaResourceManager.java:143)
at jvisa_example.lowlevel.SimpleExample.main(SimpleExample.java:75)

**Comment by pfroud** [comment]
Could you clarify about Experiment #2: only the Siglent multimeter is connected, but JVisa finds three instruments?

Have you ever used git? If you clone this repository, I can add debug printouts and push them to github so you can pull them to your computer.

**Comment by pfroud** [comment]
Sounds like we're trying to open random USB stuff which won't work.

Please replace SimpleExample.java with this file which has some debug printouts: [SimpleExample.java.txt](https://github.com/pfroud/JVisa/files/8165774/SimpleExample.java.txt)

Then run experiment number 2 only (one LAN instrument only) and reply with the output.

Git is software to keep track of changes to files. It is not easy to use so I'll just attach file to this issue. It does run on Mac (and Linux).

**Comment by Amaruk2365** [comment]
Hi Peter,


I have now replaced the code with the text file you sent me so that you can
capture more details in the outputs.


I have now run three experiments with this code. But before I show you the
results let me explain two key things that I think you need to know. First
off, on my computer I have VISA set up with three static devices as shown
in the image (my three Siglent devices). Thus, your code might think they
are available when in reality they might not be at that particular time. I
think these are the three ghost instruments your code finds.


The second thing, probably irrelevant but I mention it anyway, I have two
wifi networks but only connect to one at the time (see details in the
experiments below). When I run measurements I connect to my measurement
network router. It is not connected to the internet but my computer
connects to its network as does the LAN enabled instruments. Naturally, my
computer also has access to my regular wifi network which has access to the
internet. This is my default network I connect to when I am not performing
measurements. Also, as you will see in the first test, I tested with a more
modern Keithley SMU (2400) that has GPIB and understands the *IDN? command.
The output strings you got yesterday from the old Keithley 236 units is
what I also get when I try the *IDN? commands with them. So that was
expected and showed that your code could talk to them.


I hope this helps!


Best regards,

Mikael


=====================================

Test 1:

Connected instrument: Keithley 2400 SMU on GPIB address 11 using a USB to
GPIB adapter (NI).

LAN Status: Regular wifi network (no instruments on this network)


Output:

Found 5 VISA instrument(s):

Here are the resource names I found:

1 / 5: "GPIB0::INTFC"

2 / 5: "GPIB0::11::INSTR"

3 / 5: "TCPIP0::10.0.1.21::inst0::INSTR"

4 / 5: "TCPIP0::10.0.1.22::inst0::INSTR"

5 / 5: "TCPIP0::10.0.1.25::inst0::INSTR"

Trying to open this resource: GPIB0::INTFC

Succsessfully opened GPIB0::INTFC

Trying to open this resource: GPIB0::11::INSTR

Failed to open resource "GPIB0::INTFC": jvisa.JVisaException: in viWrite():
Unable to start operation because setup is invalid (due to attributes being
set to an inconsistent state). (0xBFFF003A)

Succsessfully opened GPIB0::11::INSTR

2 / 5: "KEITHLEY INSTRUMENTS INC.,MODEL 2400,4073783,C32   Oct  4 2010
14:20:11/A02  /U/K"

Trying to open this resource: TCPIP0::10.0.1.21::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.21::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)

Trying to open this resource: TCPIP0::10.0.1.22::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.22::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)

Trying to open this resource: TCPIP0::10.0.1.25::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.25::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)


Process finished with exit code 0

=====================================

Test 2:

Connected instrument: Siglent SDM 3045X on LAN 10.0.1.21.

LAN Status: Instrument LAN (only instruments on this network)


Output:

Found 3 VISA instrument(s):

Here are the resource names I found:

1 / 3: "TCPIP0::10.0.1.21::inst0::INSTR"

2 / 3: "TCPIP0::10.0.1.22::inst0::INSTR"

3 / 3: "TCPIP0::10.0.1.25::inst0::INSTR"

Trying to open this resource: TCPIP0::10.0.1.21::inst0::INSTR

Succsessfully opened TCPIP0::10.0.1.21::inst0::INSTR

1 / 3: "Siglent Technologies,SDM3045X,SDM34FBD4R1838,5.01.01.06R1"

Trying to open this resource: TCPIP0::10.0.1.22::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.22::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)

Trying to open this resource: TCPIP0::10.0.1.25::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.25::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)


Process finished with exit code 0

=====================================

Test 3:

Connected instrument: Nothing

LAN Status: Instrument LAN (only instruments on this network)


Output:

Found 3 VISA instrument(s):

Here are the resource names I found:

1 / 3: "TCPIP0::10.0.1.21::inst0::INSTR"

2 / 3: "TCPIP0::10.0.1.22::inst0::INSTR"

3 / 3: "TCPIP0::10.0.1.25::inst0::INSTR"

Trying to open this resource: TCPIP0::10.0.1.21::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.21::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)

Trying to open this resource: TCPIP0::10.0.1.22::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.22::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)

Trying to open this resource: TCPIP0::10.0.1.25::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.25::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)


Process finished with exit code 0


=====================================

Test 3:

Connected instrument: Nothing

LAN Status: Regular wifi network (no instruments on this network)


Output:

Found 3 VISA instrument(s):

Here are the resource names I found:

1 / 3: "TCPIP0::10.0.1.21::inst0::INSTR"

2 / 3: "TCPIP0::10.0.1.22::inst0::INSTR"

3 / 3: "TCPIP0::10.0.1.25::inst0::INSTR"

Trying to open this resource: TCPIP0::10.0.1.21::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.21::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)

Trying to open this resource: TCPIP0::10.0.1.22::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.22::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)

Trying to open this resource: TCPIP0::10.0.1.25::inst0::INSTR

Failed to open resource "TCPIP0::10.0.1.25::inst0::INSTR":
jvisa.JVisaException: in viOpen(): Insufficient location information or the
device or resource is not present in the system. (0xBFFF0011)


Process finished with exit code 0


[image: VISA setup.png]

**Comment by pfroud** [comment]
Hi Mikael, thanks for running the experiments and sharing more details about your setup.

I think you are right about the ghost instruments, and I think that means JVisa is working correctly on your computer now.

The SimpleExample file tries to open everything it can find, including static ones. When you do have instruments connected, JVisa can open them and the `*IDN?` query works.

I will publish a new release on GitHub this week.

**Comment by Amaruk2365** [comment]
Excellent! Let me know when it is up and I will download the new jar file
for my project. I am glad you figured it out!!!
You should add that your library has been verified to be MacOS compatible.

Thanks again!

Best regards,
Mikael

**Comment by pfroud** [comment]
I released JVisa-2022-03-02, please test it...

https://github.com/pfroud/JVisa/releases/tag/JVisa-2022-03-02

**Comment by Amaruk2365** [comment]
Hi Peter,
It works like a charm. I ran it with only the Siglent connected to the LAN
network and the result I got with your new jar file and the SimpleExample
code from GitHub is shown below.

Thanks again!

Best regards,
Mikael

Found 3 VISA resource(s).

Trying resource 1 / 3
   Resource name: "TCPIP0::10.0.1.21::inst0::INSTR"
   Identification query returned: "Siglent
Technologies,SDM3045X,SDM34FBD4R1838,5.01.01.06R1"

Trying resource 2 / 3
   Resource name: "TCPIP0::10.0.1.22::inst0::INSTR"
   Failed to open the resource: in viOpen(): Insufficient location
information or the device or resource is not present in the system.
(0xBFFF0011)

Trying resource 3 / 3
   Resource name: "TCPIP0::10.0.1.25::inst0::INSTR"
   Failed to open the resource: in viOpen(): Insufficient location
information or the device or resource is not present in the system.
(0xBFFF0011)

Process finished with exit code 0

**Comment by pfroud** [comment]
Great news! I will close this GitHub issue, thanks for helping me add that feature to the project.

If you can please try JVisa on Linux too and then post a comment in this issue: https://github.com/pfroud/JVisa/issues/2

**Comment by Amaruk2365** [comment]
Hi Peter,

Yes, I will try it on Linux on the weekend and let you know how it goes (I
will have to install an IDE first). No problem!

I have now also successfully written my own test program using JVisa and
talked to all my instruments (Siglent, Keysight, and Keithley) over both
GPIB and LAN. Works flawlessly!

Just a fun side-note: My Linux computer is actually a Macbook Pro that I
wiped clean and installed Linux on. Works great so I look forward to
running this experiment on it. I will also give you the details about the
Linux version I am using for your record.

Best regards,
Mikael

**Comment by Amaruk2365** [comment]
Hi again Peter,
I could not wait so I got started installing things on my Linux. IntelliJ
installed just fine. But it turns out LabVIEW (has VISA) is not supported
on my Linux version and installing the VISA by itself seems complicated
(might work but not sure). Too bad, I was hoping to test it for you but I
need to solve the issue with the VISA installation first. I might ask on
the NI forum but there is not guarantee it will work on my Linux version.

Best regards,
Mikael

**Comment by pfroud** [comment]
Bummer, thanks for trying though.

I also looked into installing NI-VISA on Linux. Their instructions are not great. If you have Ubuntu version 18.04 or 20.04 I think I know how.

Here's the link to documentation for reference:

[Download and Install NI Driver Software on Linux Desktop](https://www.ni.com/en-us/support/documentation/supplemental/18/downloading-and-installing-ni-driver-software-on-linux-desktop.html)

**Comment by Amaruk2365** [comment]
Hi Peter,
I am using Elementary OS Linux (similar to Ubuntu I think but not identical
so I can see it being an issue). At the time I read that this version was
best on Mac computers. I downloaded the VISA installation files (.deb) and
tried to follow the installation instructions given on that page. But I got
some errors (I am no Linux pro) so I have to figure out what the problem
is. LabVIEW was more clear-cut-it said it did not support it during the
installation. But the .deb files had some issues when I installed them so I
am not sure if there is something I am doing wrong. Will try some more and
if I figure it out I will let you know. It would be really great if I could
hae a Linux version to control my bench too so I will continue to look into
it.
Best regards,
Mikael
